### ISSUE TITLE
Support for built-in constructor objects

### ISSUE THREAD

**Issue by chharvey** [compat-table/compat-table#1670]
In the interest of completion, rows should be added for the built-in constructors
- [`AsyncFunction`](https://tc39.es/ecma262/#sec-async-function-constructor)
- [`GeneratorFunction`](https://tc39.es/ecma262/#sec-generatorfunction-constructor)
- [`AsyncGeneratorFunction`](https://tc39.es/ecma262/#sec-asyncgeneratorfunction-constructor)

The specification expects them to be real objects.
> The AsyncFunction constructor:
> - is %AsyncFunction%.
> - is a subclass of `Function`.
> - creates and initializes a new AsyncFunction object when called as a function rather than as a constructor…
> - is designed to be subclassable. It may be used as the value of an `extends` clause of a class definition…

So according to the spec, one should be able to write the following code:
```js
(async function () {}) instanceof AsyncFunction;

new AsyncFunction('a', 'b', 'a + b');

class MyAsynchroFn extends AsyncFunction {}

console.log(AsyncFunction.prototype);
```
and similarly for `GeneratorFunction` and `AsyncGeneratorFunction`.

Because not many (if not at all) implementations support these, it would be nice to have them on the compatibility table, if only to show which implementations are non-conforming.

**Comment by ljharb** [comment]
None of these are supposed to be globally available, and I'm not aware of any implementation that supports the syntax but not the constructors. Can you elaborate on which implementations you have in mind that are non-conforming?

**Comment by chharvey** [comment]
Can you elaborate on “None of these are supposed to be globally available”? As far as I understand it, the specification indicates that `AsyncFunction`, `GeneratorFunction`, and `AsyncGeneratorFunction` are supposed to be built-in objects and follow the rules bulleted in my block-quote above, and one should be able to reference them in code. See the [`AsyncFunction` constructor](https://tc39.es/ecma262/#sec-async-function-constructor-arguments) for example.

There are no conforming implementations as far as I can tell. In all my browsers, and in my NodeJS runtime, attempting to `console.log(AsyncFunction)` results in a ReferenceError. On [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncFunction), the pages say “Note that [this object] is not a global object” (despite them being listed under [Standard Built-In Objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects)).

So basically, even though no major implementations are conforming, it still would be nice to explicitly list these constructors on the compat table since they are technically in the ECMA specification. (And who knows, maybe *some* implementations are conforming, which is exactly what the table is for!)

**Comment by ljharb** [comment]
@chharvey the specification in fact does not make them referenceable in code directly; for example, to get to `AsyncFunction`, you have to do `(async function () {}).constructor`.

It is *quite* explicit in the spec that these things are not globally available.

**Comment by chharvey** [comment]
@ljharb  Sorry I must have missed that. Can you provide a link?

**Comment by ljharb** [comment]
There's no link to the sections in https://tc39.es/ecma262/#sec-constructor-properties-of-the-global-object that simply don't exist :-)

**Comment by chharvey** [comment]
Ah okay. I was unaware that they needed to be properties of the Global Object. The way that each section is worded, it leads me to believe that they can just be accessed (e.g. `console.log(AsyncFunction)`). Thanks for the link and for clearing that up.

If I could offer a suggestion, it would be to provide an informal note that explicitly lists the constructors that are defined in the spec but that *aren’t* properties of the Global Object. Just so readers know they haven’t been “accidentally left out” as it were.
